# <center> Web Scraping </center> 

In [1]:
from requests import get
from bs4 import BeautifulSoup

In [1]:
import tmdbsimple as tmdb

In [2]:
tmdb.API_KEY = 'f875e3c0cde708e575e3b72bea080a66'

## 1. Beautiful Soup:

Using Python's Beautiful Soup to get data from IMDB's Top 150 movies

In [2]:
url = 'http://www.imdb.com/search/title?release_date=2017&sort=num_votes,desc&page=1'

In [3]:
response = get(url)
print(response.text[:500])



<!DOCTYPE html>
<html
    xmlns:og="http://ogp.me/ns#"
    xmlns:fb="http://www.facebook.com/2008/fbml">
    <head>
         
        <meta charset="utf-8">
        <meta http-equiv="X-UA-Compatible" content="IE=edge">

    <meta name="apple-itunes-app" content="app-id=342792525, app-argument=imdb:///?src=mdot">



        <script type="text/javascript">var IMDbTimer={starttime: new Date().getTime(),pt:'java'};</script>

<script>
    if (typeof uet == 'function') {
      uet("bb", "LoadTitle",


In [47]:
html_soup = BeautifulSoup(response.text, 'html.parser')
type(html_soup)

bs4.BeautifulSoup

In [48]:
movie_containers = html_soup.find_all('div', class_ = 'lister-item mode-advanced')
print(type(movie_containers))
print(len(movie_containers))

<class 'bs4.element.ResultSet'>
50


In [49]:
first_movie = movie_containers[0]
first_movie

<div class="lister-item mode-advanced">
<div class="lister-top-right">
<div class="ribbonize" data-caller="filmosearch" data-tconst="tt3315342"></div>
</div>
<div class="lister-item-image float-left">
<a href="/title/tt3315342/?ref_=adv_li_i"> <img alt="Logan" class="loadlate" data-tconst="tt3315342" height="98" loadlate="https://m.media-amazon.com/images/M/MV5BYzc5MTU4N2EtYTkyMi00NjdhLTg3NWEtMTY4OTEyMzJhZTAzXkEyXkFqcGdeQXVyNjc1NTYyMjg@._V1_UX67_CR0,0,67,98_AL_.jpg" src="https://m.media-amazon.com/images/G/01/imdb/images/nopicture/large/film-184890147._CB470041630_.png" width="67"/>
</a> </div>
<div class="lister-item-content">
<h3 class="lister-item-header">
<span class="lister-item-index unbold text-primary">1.</span>
<a href="/title/tt3315342/?ref_=adv_li_tt">Logan</a>
<span class="lister-item-year text-muted unbold">(2017)</span>
</h3>
<p class="text-muted ">
<span class="certificate">R</span>
<span class="ghost">|</span>
<span class="runtime">137 min</span>
<span class="ghost">|</

In [8]:
first_movie.div

<div class="lister-top-right">
<div class="ribbonize" data-caller="filmosearch" data-tconst="tt3315342"></div>
</div>

In [9]:
first_movie.a

<a href="/title/tt3315342/?ref_=adv_li_i"> <img alt="Logan" class="loadlate" data-tconst="tt3315342" height="98" loadlate="https://m.media-amazon.com/images/M/MV5BYzc5MTU4N2EtYTkyMi00NjdhLTg3NWEtMTY4OTEyMzJhZTAzXkEyXkFqcGdeQXVyNjc1NTYyMjg@._V1_UX67_CR0,0,67,98_AL_.jpg" src="https://m.media-amazon.com/images/G/01/imdb/images/nopicture/large/film-184890147._CB470041630_.png" width="67"/>
</a>

In [10]:
first_movie.h3

<h3 class="lister-item-header">
<span class="lister-item-index unbold text-primary">1.</span>
<a href="/title/tt3315342/?ref_=adv_li_tt">Logan</a>
<span class="lister-item-year text-muted unbold">(2017)</span>
</h3>

In [11]:
first_movie.h3.a

<a href="/title/tt3315342/?ref_=adv_li_tt">Logan</a>

In [12]:
first_name = first_movie.h3.a.text
first_name

'Logan'

In [13]:
first_year = first_movie.h3.find('span', class_ = 'lister-item-year text-muted unbold').text
first_year

'(2017)'

In [14]:
first_imdb = float(first_movie.strong.text)
first_imdb

8.1

In [15]:
first_mscore = first_movie.find('span', class_ = 'metascore favorable')

first_mscore = int(first_mscore.text)
print(first_mscore)

77


In [17]:
first_votes = first_movie.find('span', attrs = {'name':'nv'})['data-value']
first_votes

'523398'

In [20]:
type(first_votes)

str

In [22]:
first_votes = int(first_votes)

In [57]:
first_runtime = first_movie.p.find('span', {'class': 'runtime'}).text
first_runtime

'137 min'

In [79]:
first_movie.find_all('p')[1].get_text()

"\n    In the near future, a weary Logan cares for an ailing Professor X. However, Logan's attempts to hide from the world and his legacy are upended when a young mutant arrives, pursued by dark forces."

In [74]:
first_text = first_movie.select("p:nth-of-type(2)")[0].get_text()
first_text

"\n    In the near future, a weary Logan cares for an ailing Professor X. However, Logan's attempts to hide from the world and his legacy are upended when a young mutant arrives, pursued by dark forces."

In [83]:
first_movie.find_all('p')[2]

<p class="">
    Director:
<a href="/name/nm0003506/?ref_=adv_li_dr_0">James Mangold</a>
<span class="ghost">|</span> 
    Stars:
<a href="/name/nm0413168/?ref_=adv_li_st_0">Hugh Jackman</a>, 
<a href="/name/nm0001772/?ref_=adv_li_st_1">Patrick Stewart</a>, 
<a href="/name/nm6748436/?ref_=adv_li_st_2">Dafne Keen</a>, 
<a href="/name/nm2933542/?ref_=adv_li_st_3">Boyd Holbrook</a>
</p>

In [88]:
first_director = first_movie.find_all('p')[2].a.get_text()
first_director

'James Mangold'

In [108]:
temp_first_cc = []

In [109]:
first_movie.find_all('p')[2].find_all('a')

[<a href="/name/nm0003506/?ref_=adv_li_dr_0">James Mangold</a>,
 <a href="/name/nm0413168/?ref_=adv_li_st_0">Hugh Jackman</a>,
 <a href="/name/nm0001772/?ref_=adv_li_st_1">Patrick Stewart</a>,
 <a href="/name/nm6748436/?ref_=adv_li_st_2">Dafne Keen</a>,
 <a href="/name/nm2933542/?ref_=adv_li_st_3">Boyd Holbrook</a>]

In [110]:
for tag in first_movie.find_all('p')[2].find_all('a'):
    temp_first_cc.append(tag.text)

In [111]:
temp_first_cc

['James Mangold',
 'Hugh Jackman',
 'Patrick Stewart',
 'Dafne Keen',
 'Boyd Holbrook']

In [112]:
first_cc = ", ".join(temp_first_cc)

In [113]:
first_cc

'James Mangold, Hugh Jackman, Patrick Stewart, Dafne Keen, Boyd Holbrook'

In [115]:
print("First Movie Desc -- ", first_movie)

First Movie Desc --  <div class="lister-item mode-advanced">
<div class="lister-top-right">
<div class="ribbonize" data-caller="filmosearch" data-tconst="tt3315342"></div>
</div>
<div class="lister-item-image float-left">
<a href="/title/tt3315342/?ref_=adv_li_i"> <img alt="Logan" class="loadlate" data-tconst="tt3315342" height="98" loadlate="https://m.media-amazon.com/images/M/MV5BYzc5MTU4N2EtYTkyMi00NjdhLTg3NWEtMTY4OTEyMzJhZTAzXkEyXkFqcGdeQXVyNjc1NTYyMjg@._V1_UX67_CR0,0,67,98_AL_.jpg" src="https://m.media-amazon.com/images/G/01/imdb/images/nopicture/large/film-184890147._CB470041630_.png" width="67"/>
</a> </div>
<div class="lister-item-content">
<h3 class="lister-item-header">
<span class="lister-item-index unbold text-primary">1.</span>
<a href="/title/tt3315342/?ref_=adv_li_tt">Logan</a>
<span class="lister-item-year text-muted unbold">(2017)</span>
</h3>
<p class="text-muted ">
<span class="certificate">R</span>
<span class="ghost">|</span>
<span class="runtime">137 min</span>
<s

In [117]:
print("First Movie Name -- ", first_name)
print("----------------------------")
print("First Movie Year -- ", first_year)
print("----------------------------")
print("First Movie IMDV -- ", first_imdb)
print("----------------------------")
print("First Movie MSCR -- ", first_mscore)
print("----------------------------")
print("First Movie Votes -- ", first_votes)
print("----------------------------")
print("First Movie Runtime -- ", first_runtime)
print("----------------------------")
print("First Movie Director -- ", first_director)
print("----------------------------")
print("First Movie Cast and Crew -- ", first_cc)
print("----------------------------")

First Movie Name --  Logan
----------------------------
First Movie Year --  (2017)
----------------------------
First Movie IMDV --  8.1
----------------------------
First Movie MSCR --  77
----------------------------
First Movie Votes --  523398
----------------------------
First Movie Runtime --  137 min
----------------------------
First Movie Director --  James Mangold
----------------------------
First Movie Cast and Crew --  James Mangold, Hugh Jackman, Patrick Stewart, Dafne Keen, Boyd Holbrook
----------------------------


## 2. Tmdbsimple

### (i) Using tmdb.Movies()

In [1]:
import tmdbsimple as tmdb

In [2]:
tmdb.API_KEY = 'f875e3c0cde708e575e3b72bea080a66'

In [3]:
movie = tmdb.Movies(603)

In [4]:
movie

In [10]:
movie = movie.info()

In [12]:
movie

{'adult': False,
 'backdrop_path': '/7u3pxc0K1wx32IleAkLv78MKgrw.jpg',
 'belongs_to_collection': {'backdrop_path': '/bRm2DEgUiYciDw3myHuYFInD7la.jpg',
  'id': 2344,
  'name': 'The Matrix Collection',
  'poster_path': '/lh4aGpd3U9rm9B8Oqr6CUgQLtZL.jpg'},
 'budget': 63000000,
 'genres': [{'id': 28, 'name': 'Action'},
  {'id': 878, 'name': 'Science Fiction'}],
 'homepage': 'http://www.warnerbros.com/matrix',
 'id': 603,
 'imdb_id': 'tt0133093',
 'original_language': 'en',
 'original_title': 'The Matrix',
 'overview': 'Set in the 22nd century, The Matrix tells the story of a computer hacker who joins a group of underground insurgents fighting the vast and powerful computers who now rule the earth.',
 'popularity': 29.877,
 'poster_path': '/hEpWvX6Bp79eLxY1kX5ZZJcme5U.jpg',
 'production_companies': [{'id': 79,
   'logo_path': '/tpFpsqbleCzEE2p5EgvUq6ozfCA.png',
   'name': 'Village Roadshow Pictures',
   'origin_country': 'US'},
  {'id': 372,
   'logo_path': None,
   'name': 'Groucho II Film

In [15]:
movie['adult']

False

In [16]:
movie['title']

'The Matrix'

In [17]:
movie['budget']

63000000

In [18]:
movie['overview']

'Set in the 22nd century, The Matrix tells the story of a computer hacker who joins a group of underground insurgents fighting the vast and powerful computers who now rule the earth.'

In [19]:
movie['revenue']

463517383

In [20]:
movie['runtime']

136

In [217]:
movie_1 = tmdb.Movies('The Dark Knight')

In [219]:
movie_1.info()

HTTPError: 404 Client Error: Not Found for url: https://api.themoviedb.org/3/movie/The%20Dark%20Knight?api_key=f875e3c0cde708e575e3b72bea080a66

** Note:** We can query TMDB API only using movie_ids and not by movie titles. (When queried, API throws a 404 Cleint Error)
Also, quering with titles take singnificantly longer time to that of movie_id.

### (ii) Using tmdb.search()

In [35]:
search = tmdb.Search()
response = search.movie(query='The Bourne')

In [36]:
for s in search.results:
    print(s['title'], s['id'], s['release_date'], s['popularity'])

The Bourne Identity 2501 2002-06-14 13.959
The Bourne Supremacy 2502 2004-07-23 13.047
The Bourne Legacy 49040 2012-08-08 12.672
The Bourne Ultimatum 2503 2007-08-03 12.482
Bette Bourne: It Goes with the Shoes 179304 2013-03-21 0.6
Jason Bourne 324668 2016-07-27 13.083
Untitled Jeremy Renner/Bourne Sequel 393640  0.806


In [39]:
search = tmdb.Search()
response = search.movie(query='The Dark Knight')

In [40]:
for s in search.results:
    print(s['title'], s['id'], s['release_date'], s['popularity'])

The Dark Knight 155 2008-07-16 26.17
The Dark Knight Rises 49026 2012-07-16 24.649
Batman: The Dark Knight Returns, Part 2 142061 2013-01-18 8.58
Batman: The Dark Knight Returns, Part 1 123025 2012-09-06 8.051
Garo - Kiba: The Dark Knight 257171 2011-09-03 1.072
Batman: The Dark Knight Returns (Deluxe Edition) 472027 2013-10-08 1.79
Batman Unmasked: The Psychology of the Dark Knight 29751 2008-07-15 1.78
The Dark Knight 72003 2011-07-11 0.6
Legends of the Dark Knight: The History of Batman 141559 2005-11-15 1.059
Batman: The Dark Knight's First Night 377531 1991-01-12 0.851
Batman: The Animated Series -  Tales of the Dark Knight 501972  0.6
Masterpiece: Frank Miller's The Dark Knight Returns 369661 2013-10-08 0.6
Shadows of the Bat: The Cinematic Saga of the Dark Knight 204651 2005-10-18 0.775
The Dark Knight Returns: An Epic Fan Film 458369 2016-11-07 0.6
The Fire Rises: The Creation and Impact of The Dark Knight Trilogy 243238 2013-09-24 0.672
Shadows of the Bat: The Cinematic Saga o

In [45]:
search.results[0]

{'adult': False,
 'backdrop_path': '/hqkIcbrOHL86UncnHIsHVcVmzue.jpg',
 'genre_ids': [18, 28, 80, 53],
 'id': 155,
 'original_language': 'en',
 'original_title': 'The Dark Knight',
 'overview': 'Batman raises the stakes in his war on crime. With the help of Lt. Jim Gordon and District Attorney Harvey Dent, Batman sets out to dismantle the remaining criminal organizations that plague the streets. The partnership proves to be effective, but they soon find themselves prey to a reign of chaos unleashed by a rising criminal mastermind known to the terrified citizens of Gotham as the Joker.',
 'popularity': 26.17,
 'poster_path': '/1hRoyzDtpgMU7Dz4JF22RANzQO7.jpg',
 'release_date': '2008-07-16',
 'title': 'The Dark Knight',
 'video': False,
 'vote_average': 8.4,
 'vote_count': 16998}

In [49]:
print(search.results[0]['id'], search.results[0]['title'], 
search.results[0]['overview'])

155 The Dark Knight Batman raises the stakes in his war on crime. With the help of Lt. Jim Gordon and District Attorney Harvey Dent, Batman sets out to dismantle the remaining criminal organizations that plague the streets. The partnership proves to be effective, but they soon find themselves prey to a reign of chaos unleashed by a rising criminal mastermind known to the terrified citizens of Gotham as the Joker.


** Note: ** When we use tmdb.search() we do get the tmdb_id as well as the title. But using tmdb.Movies() yields much more information about the movie.

** New Approach:** 

We can query TMDB API only using movie_ids and not by movie titles. (When queried, API throws a 404 Cleint Error)
Also, quering with titles take singnificantly longer time to that of movie_id.

However, Movie Lens dataset has its own movie_id which are quite different from that of TMDBs (tmdb_id)

Hence, we use the follwoing approach:

- Get the movielens_id and title from movielens dataset
- Query TMDB API using movie title to get TMDB_IDs
- Use queried tmdb_id to get additional info about the movie 

### (iii) Movie Lens Data

In [52]:
# Jupyter Notebook with Matplotlib Inline
%matplotlib notebook

# Importing necessary modules
import os
import pandas as pd
import numpy as np
import seaborn as sns
import matplotlib as plt
import matplotlib.pyplot as plt
from sklearn.model_selection import train_test_split
from sklearn import metrics
from subprocess import check_output

from sklearn.metrics import pairwise_distances

In [53]:
# Set height, width, maximum rows and columns
pd.set_option('display.max_rows', 500)
pd.set_option('display.max_columns', 500)
pd.set_option('display.width', 1000)

In [50]:
u_cols = ['user_id', 'age', 'sex', 'occupation', 'zip_code']
r_cols = ['user_id', 'movie_id', 'rating', 'timestamp']
m_cols = ['movie_id', 'title', 'release_date', 'video_release_date', 'imdb_url','unknown', 'Action', 'Adventure',
'Animation', 'Children\'s', 'Comedy', 'Crime', 'Documentary', 'Drama', 'Fantasy','Film-Noir', 'Horror', 'Musical', 'Mystery', 'Romance', 'Sci-Fi', 'Thriller', 'War', 'Western']

In [55]:
users_1 = pd.read_csv("data/u.user",sep='|',names=u_cols)
ratings_1 = pd.read_csv('data/u.data',sep='\t', names=r_cols)
movies_1 = pd.read_csv('data/u.item', sep='|', names=m_cols, encoding='latin-1')
movielens=pd.merge(users_1 , ratings_1)
movielens=pd.merge(movielens,movies_1)

In [56]:
movielens.head(3)

user_id  age sex  occupation zip_code  movie_id  rating  timestamp                       title release_date  video_release_date                                           imdb_url  unknown  Action  Adventure  Animation  Children's  Comedy  Crime  Documentary  Drama  Fantasy  Film-Noir  Horror  Musical  Mystery  Romance  Sci-Fi  Thriller  War  Western
0        1   24   M  technician    85711        61       4  878542420  Three Colors: White (1994)  01-Jan-1994                 NaN  http://us.imdb.com/M/title-exact?Trzy%20kolory...        0       0          0          0           0       0      0            0      1        0          0       0        0        0        0       0         0    0        0
1       13   47   M    educator    29206        61       4  882140552  Three Colors: White (1994)  01-Jan-1994                 NaN  http://us.imdb.com/M/title-exact?Trzy%20kolory...        0       0          0          0           0       0      0            0      1        0          0       0        0        0        0       0         0    0        0
2       18   35   F       other    37212        61       4  880130803  Three Colors: White (1994)  01-Jan-1994                 NaN  http://us.imdb.com/M/title-exact?Trzy%20kolory...        0       0          0          0           0       0      0            0      1        0          0       0        0        0        0       0         0    0        0

In [109]:
td = movielens.copy(deep = True)

In [110]:
# Format 'title' i.e. remove 'year' from title
td['title'] = td['title'].astype(str).str[:-7]

In [113]:
td = td[['movie_id', 'rating', 'title']]

In [124]:
new_td = td.groupby(['movie_id', 'title']).size().reset_index()

In [125]:
new_td.head()

movie_id       title    0
0         1   Toy Story  452
1         2   GoldenEye  131
2         3  Four Rooms   90
3         4  Get Shorty  209
4         5     Copycat   86

In [126]:
new_td['title'].head()

0     Toy Story
1     GoldenEye
2    Four Rooms
3    Get Shorty
4       Copycat
Name: title, dtype: object

### (iv) Putting it all together

In [339]:
# Movies array
mvs = np.array(new_td['title'])

In [220]:
import copy
tmdb_ids = copy.deepcopy(new_omdb_id)
tmdb_titles = copy.deepcopy(new_omdb_titles)

In [340]:
# Omdb Id and Title arrays
tmdb_ids = []
tmdb_titles = []

In [341]:
for mv in mvs:                                            
    if (mv != ''):                                        # If movie name is not null
        search = tmdb.Search()                            # Initialize a search object
        response = search.movie(query = mv)               # Search for the movie name

        if (response['results'] != []):                   # If the search results are not null
            tmdb_ids.append(search.results[0]['id'])      # Pick the closely matched entry and append id to tmdb_ids
            tmdb_titles.append(search.results[0]['title'])# Pick the closely matched entry and append id to tmdb_titles
        else:
            continue

In [342]:
# Check if tmdb ids length and tmdb titles length are equal
len(tmdb_ids) == len(tmdb_titles)

True

In [343]:
# Number of movies that are in movie lens dataset but not in the omdb
len(mvs) - len(tmdb_titles)

64

In [344]:
# List to save the data queried using tmdb.Movies()
d = []

In [346]:
# Gathering additional data using tmdb.Movies()
for t_id in tmdb_ids:
    movie = tmdb.Movies(t_id)
    movie = movie.info()    
    d.append((movie['id'], movie['title'], movie['adult'], movie['budget'], movie['overview'], movie['revenue'], 
              movie['runtime'], movie['vote_average'], movie['vote_count']))

In [347]:
# Converting the list to a dataframe
tmdb_df = pd.DataFrame(d, columns=('tmdb_id', 'title', 'adult', 'budget', 'overview', 'revenue', 'runtime', 'imdb_avg_rating', 'imdb_rating_count'))

In [348]:
tmdb_df.head(2)

tmdb_id      title  adult    budget                                           overview    revenue  runtime  imdb_avg_rating  imdb_rating_count
0      862  Toy Story  False  30000000  Led by Woody, Andy's toys live happily in his ...  373554033     81.0              7.9               8883
1      710  GoldenEye  False  58000000  James Bond must unmask the mysterious head of ...  352194034    130.0              6.8               1761

In [228]:
tmdb_df.shape

(1618, 9)

In [229]:
# Verifying if length of new dataframe is same as that of tmdb_ids
tmdb_df.shape[0] == len(tmdb_ids)

True

### (v) Movies that did not match the results

In [249]:
missing_titles = []

In [250]:
for mv in mvs:                                            
    if (mv == ''):                                        # If movie name is not null
        missing_titles.append(mv)
    
    else:
        search = tmdb.Search()                            # Initialize a search object
        response = search.movie(query = mv)               # Search for the movie name

        if (response['results'] == []):                   # If the search results are not null
            missing_titles.append(mv)
        else:
            continue

In [251]:
len(missing_titles)

64

In [252]:
missing_titles

['Shanghai Triad (Yao a yao yao dao waipo qiao)',
 'Horseman on the Roof, The (Hussard sur le toit, Le)',
 'Jungle2Jungle',
 'When the Cats Away (Chacun cherche son chat)',
 '',
 'My Life as a Dog (Mitt liv som hund)',
 "Stephen King's The Langoliers",
 'Police Story 4: Project S (Chao ji ji hua)',
 "Daniel Defoe's Robinson Crusoe",
 'Tin Drum, The (Blechtrommel, Die)',
 'Seventh Seal, The (Sjunde inseglet, Det)',
 "Blood For Dracula (Andy Warhol's Dracula)",
 'Queen Margot (Reine Margot, La)',
 'Ma vie en rose (My Life in Pink)',
 'U.S. Marshalls',
 'Flower of My Secret, The (Flor de mi secreto, La)',
 'Until the End of the World (Bis ans Ende der Welt)',
 'Die xue shuang xiong (Killer, The)',
 "Star Maker, The (Uomo delle stelle, L')",
 'Umbrellas of Cherbourg, The (Parapluies de Cherbourg, Les)',
 'Fille seule, La (A Single Girl)',
 'Before the Rain (Pred dozhdot)',
 'Savage Nights (Nuits fauves, Les)',
 'Marlene Dietrich: Shadow and Light ',
 'Maybe, Maybe Not (Bewegte Mann, Der)',

In [241]:
new_tmdb_id = list(tmdb_df['tmdb_id'])

In [243]:
len(new_tmdb_id)

1618

** Note:** Manually checking the movies, I see that if movies quried excluding the parenthesis area, do return results from TMDB API. Also, the movie titles that have 'The' as a prefix have a different notation in the Movie Lens dataset

<strong> Example: </strong> <br>
Actual Movie Title - 'The Dark Knight' <br>
Movie Lens Title - 'Dark Knight, The'

TMDB API considers Actual Movie Title. To handle this we have to pre-process our movie titles by removing the parenthesis area as well as handle the suffixes in the movie title.

In [262]:
search = tmdb.Search()
response = search.movie(query='Land and Freedom')

In [263]:
for s in search.results:
    print(s['title'], s['id'], s['release_date'], s['popularity'])

Land and Freedom 38884 1995-01-01 3.714
Land of Truth, Love & Freedom 328362 2000-11-11 0.631


In [273]:
# Removing parenthesis area in a string
import re
items = copy.deepcopy(missing_titles)
filtered_missing_titles = []
for item in items:
    filtered_missing_titles.append(re.sub(r" ?\([^)]+\)", "", item))

In [282]:
# Handling prefixes in the movie title
new_missing_titles = []
for i in filtered_missing_titles:
    if ',' in i:
        ls = i.split(", ")
        strng = ls[1] + " " + ls[0]
        new_missing_titles.append(strng)
    else:
        new_missing_titles.append(i)

In [271]:
len(new_missing_titles)

64

In [283]:
new_missing_titles

['Shanghai Triad',
 'The Horseman on the Roof',
 'Jungle2Jungle',
 'When the Cats Away',
 '',
 'My Life as a Dog',
 "Stephen King's The Langoliers",
 'Police Story 4: Project S',
 "Daniel Defoe's Robinson Crusoe",
 'The Tin Drum',
 'The Seventh Seal',
 'Blood For Dracula',
 'Queen Margot',
 'Ma vie en rose',
 'U.S. Marshalls',
 'The Flower of My Secret',
 'Until the End of the World',
 'Die xue shuang xiong',
 'The Star Maker',
 'The Umbrellas of Cherbourg',
 'La Fille seule',
 'Before the Rain',
 'Savage Nights',
 'Marlene Dietrich: Shadow and Light ',
 'Maybe Not Maybe',
 'The Big Bang Theory',
 'The Old Lady Who Walked in the Sea',
 'Talking About Sex',
 'Killer',
 'Metisse',
 "The Wooden Man's Bride",
 'His Life and Music Last Klezmer: Leopold Kozlowski',
 'My Life and Times With Antonin Artaud',
 'Midnight Dancers',
 'The Day the Sun Turned Cold',
 'Shadow of Angels',
 'The Deadly Cure',
 'Boys in Venice',
 'The Forbidden Christ',
 "I Can't Sleep",
 "Roseanna's Grave",
 "Boy's Lif

### (vi) Putting it all together (again)

In [326]:
# Dataframe with original title and cleaned titles
df_zipped = pd.DataFrame(list(zip(missing_titles, new_missing_titles)), columns = ['original_title', 'cleaned_title'])

In [327]:
df_zipped.head()

original_title             cleaned_title
0      Shanghai Triad (Yao a yao yao dao waipo qiao)            Shanghai Triad
1  Horseman on the Roof, The (Hussard sur le toit...  The Horseman on the Roof
2                                      Jungle2Jungle             Jungle2Jungle
3       When the Cats Away (Chacun cherche son chat)        When the Cats Away
4

In [131]:
# New movies array
new_missing_titles

In [284]:
# Omdb Id and Title arrays
new_tmdb_ids = []
new_tmdb_titles = []

In [285]:
for mv in new_missing_titles:                                            
    if (mv != ''):                                        # If movie name is not null
        search = tmdb.Search()                            # Initialize a search object
        response = search.movie(query = mv)               # Search for the movie name

        if (response['results'] != []):                   # If the search results are not null
            new_tmdb_ids.append(search.results[0]['id'])      # Pick the closely matched entry and append id to tmdb_ids
            new_tmdb_titles.append(search.results[0]['title'])# Pick the closely matched entry and append id to tmdb_titles
        else:
            continue

In [286]:
# Check if new_tmdb_ids length and new_tmdb_titles length are equal
len(new_tmdb_ids) == len(new_tmdb_titles)

True

In [289]:
# Number of movies that are in movie lens dataset but not in the omdb
len(new_missing_titles) - len(new_tmdb_titles)

19

In [333]:
# List to save the data queried using tmdb.Movies()
d1 = []

In [334]:
# Gathering additional data using tmdb.Movies()
for t_id in new_tmdb_ids:
    movie = tmdb.Movies(t_id)
    movie = movie.info()    
    d1.append((movie['id'], movie['title'], movie['adult'], movie['budget'], movie['overview'], movie['revenue'], 
              movie['runtime'], movie['vote_average'], movie['vote_count']))

In [335]:
# Converting the list to a dataframe
tmdb_df_1 = pd.DataFrame(d1, columns=('tmdb_id', 'title', 'adult', 'budget', 'overview', 'revenue', 'runtime', 'imdb_avg_rating', 'imdb_rating_count'))

In [336]:
tmdb_df_1.head(2)

tmdb_id                     title  adult    budget                                           overview   revenue  runtime  imdb_avg_rating  imdb_rating_count
0    37557            Shanghai Triad  False   2300000  A provincial boy related to a Shanghai crime f...         0    108.0              6.9                 26
1    11876  The Horseman on the Roof  False  35000000  In a time of war and disease, a young officer ...  15000000    135.0              6.5                 48

In [337]:
tmdb_df_1.shape

(45, 9)

In [338]:
# Verifying if length of new dataframe is same as that of tmdb_ids
tmdb_df_1.shape[0] == len(new_tmdb_ids)

True

In [351]:
dfs = [tmdb_df, tmdb_df_1]

In [353]:
tmdbs = pd.concat(dfs)

In [355]:
tmdbs.shape[0]

1663

In [357]:
tmdb_df.shape[0]+tmdb_df_1.shape[0]

1663

In [354]:
tmdbs.head()

tmdb_id       title  adult    budget                                           overview    revenue  runtime  imdb_avg_rating  imdb_rating_count
0      862   Toy Story  False  30000000  Led by Woody, Andy's toys live happily in his ...  373554033     81.0              7.9               8883
1      710   GoldenEye  False  58000000  James Bond must unmask the mysterious head of ...  352194034    130.0              6.8               1761
2        5  Four Rooms  False   4000000  It's Ted the Bellhop's first night on the job....    4300000     98.0              6.3                980
3     8012  Get Shorty  False  30250000  Chili Palmer is a Miami mobster who gets sent ...  115101622    105.0              6.5                430
4     1710     Copycat  False         0  An agoraphobic psychologist and a female detec...          0    124.0              6.5                344

In [358]:
new_td.shape[0]

1682

In [370]:
tmdbs.to_csv('tmdbs.csv', sep = ',')